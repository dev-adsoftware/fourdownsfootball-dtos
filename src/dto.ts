import { ClassConstructor, plainToClass } from 'class-transformer';
import { validateSync, ValidationError } from 'class-validator';

export abstract class Dto {
  public init(obj: Record<string | number, unknown>): this {
    Object.assign(
      this,
      plainToClass(this.constructor as ClassConstructor<this>, { ...obj }),
    );
    return this.validate(true);
  }

  public serialize(): string {
    return JSON.stringify(this);
  }

  private validate(whitelist: boolean): this {
    const errors = validateSync(this, {
      whitelist,
      validationError: { target: false },
    });
    if (errors.length > 0) {
      throw new Error(
        `${this.constructor.name}: ${this.validationErrors(errors)}`,
      );
    }

    return this;
  }

  private validationErrors(errors: ValidationError[]): string {
    return errors
      .map((e) => {
        if (e.constraints) return Object.values(e.constraints).join(', ');
        return String(e);
      })
      .join(', ');
  }
}
