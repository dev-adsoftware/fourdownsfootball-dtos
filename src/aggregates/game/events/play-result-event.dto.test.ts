import { PlayResultEvent } from './play-result-event.dto';

describe('given: missing data', () => {
  describe('when: I initialize', () => {
    it('then: error matched snapshot', async () => {
      try {
        expect(new PlayResultEvent().init({})).not.toBeDefined();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });
});
