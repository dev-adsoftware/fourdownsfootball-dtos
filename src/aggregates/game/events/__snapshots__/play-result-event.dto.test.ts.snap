// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given: missing data when: I initialize then: error matched snapshot 1`] = `[Error: PlayResultEvent: attributes must be an object, nextSeed must be a number conforming to the specified constraints]`;

exports[`given: missing nested data when: I initialize then: error matched snapshot 1`] = `
[Error: PlayResultEvent: An instance of an object has failed the validation:
 - property attributes.description has failed the following constraints: isString 
 - property attributes.elapsedTime has failed the following constraints: isNumber 
 - property attributes.isOutOfBounds has failed the following constraints: isBoolean 
, nextSeed must be a number conforming to the specified constraints]
`;
